sql_generators/glean_usage/glean_app_ping_views.py

```diff
@@ -66,52 +66,153 @@ def generate_per_app(self, project_id, app_info, output_dir=None):
-                        underlying_view_id=underlying_view_id,
+                        app_channels=", ".join(app_channels),
                     ),
                     skip_existing=True,
                 )
 
-                # we create a schema to the original view created for the
-                # stable tables here (this assumes that they have been or
-                # will be generated, which should be the case for a full
-                # run of the sql generation logic)
                 schema_dir = get_table_dir(output_dir, full_view_id)
-                original_schema_file = os.path.relpath(
-                    os.path.abspath(
-                        get_table_dir(output_dir, underlying_view_id) / "schema.yaml"
-                    ),
-                    start=schema_dir,
-                )
-                schema_link = schema_dir / "schema.yaml"
-                try:
-                    os.unlink(schema_link)
-                except FileNotFoundError:
-                    pass
-                os.symlink(original_schema_file, schema_link)
+                unioned_schema.to_yaml_file(schema_dir / "schema.yaml")
+
+    def _generate_select_expression(
+        self, unioned_schema_nodes, app_schema_nodes, path=[]
+    ) -> str:
+        """
+        Generate the select expression based on the unioned schema and the app channel schema.
+
+        Any fields that are missing in the app_schema are set to NULL.
+        """
+        select_expr = []
+        unioned_schema_nodes = {n["name"]: n for n in unioned_schema_nodes}
+        app_schema_nodes = {n["name"]: n for n in app_schema_nodes}
+
+        # iterate through fields
+        for node_name, node in unioned_schema_nodes.items():
+            dtype = node["type"]
+
+            if node_name in app_schema_nodes:
+                # field exists in app schema
+
+                if node == app_schema_nodes[node_name]:
+                    # field (and all nested fields) are identical, so just query it
+                    select_expr.append(f"{'.'.join(path + [node_name])}")
+                else:
+                    # fields, and/or nested fields are not identical
+                    if dtype == "RECORD":
+                        # for nested fields, recursively generate select expression
+
+                        if node.get("mode", None) == "REPEATED":
+                            # unnest repeated record
+                            select_expr.append(f"""
+                                ARRAY_AGG(
+                                    (   SELECT
+                                        STRUCT(
+                                            {self._generate_select_expression(node['fields'], app_schema_nodes[node_name]['fields'], path + [node_name])}
+                                        )
+                                        FROM UNNEST({node_name}) AS {node_name}
+                                    )
+                                ) AS {node_name}
```

Comment:
- With the aggregate function not in a subquery the overall query will complain about the other columns not being grouped or aggregated.
- The identifier within `UNNEST()` needs to be fully qualified.
- The inner select expressions need to be qualified with the `UNNEST()` alias rather than the full path.

```suggestion
                                (
                                    SELECT ARRAY_AGG(
                                        STRUCT(
                                            {self._generate_select_expression(node['fields'], app_schema_nodes[node_name]['fields'], [node_name])}
                                        )
                                    )
                                    FROM UNNEST({'.'.join(path + [node_name])}) AS {node_name}
                                ) AS {node_name}
```

Reply:
good catch, I'll push a fix